Response-Theory/Self-Consistent-Field/beta.py

```diff
@@ -0,0 +1,351 @@
+    jk.C_left_add(L)
+    jk.C_right_add(R)
+    jk.compute()
+    # 1/2 is due to X + Y
+    J = 0.5 * np.asarray(jk.J()[0])
+    K = 0.5 * np.asarray(jk.K()[0])
+
+    # eqn. (21b)
+    F = (C.T).dot(4 * J - K.T - K).dot(C)
+    G[i, ...] = V + F
+
+# form epsilon matrices, eqn. (34)
+E = G.copy()
+omega = 0
+for i in range(ncomp):
+    eoU = (moenergies[..., na] + omega) * U[i, ...]
+    Ue = U[i, ...] * moenergies[na, ...]
+    E[i, ...] += (eoU - Ue)
+
+# Assume some symmetry and calculate only part of the tensor.
+# eqn. (VII-4)
+hyperpolarizability = np.zeros(shape=(6, 3))
+off1 = [0, 1, 2, 0, 0, 1]
+off2 = [0, 1, 2, 1, 2, 2]
+for r in range(6):
+    b = off1[r]
+    c = off2[r]
+    for a in range(3):
+        tl1 = 2 * np.trace(U[a, ...].dot(G[b, ...]).dot(U[c, ...])[:nocc, :nocc])
+        tl2 = 2 * np.trace(U[a, ...].dot(G[c, ...]).dot(U[b, ...])[:nocc, :nocc])
+        tl3 = 2 * np.trace(U[c, ...].dot(G[a, ...]).dot(U[b, ...])[:nocc, :nocc])
+        tr1 = np.trace(U[c, ...].dot(U[b, ...]).dot(E[a, ...])[:nocc, :nocc])
+        tr2 = np.trace(U[b, ...].dot(U[c, ...]).dot(E[a, ...])[:nocc, :nocc])
+        tr3 = np.trace(U[c, ...].dot(U[a, ...]).dot(E[b, ...])[:nocc, :nocc])
+        tr4 = np.trace(U[a, ...].dot(U[c, ...]).dot(E[b, ...])[:nocc, :nocc])
+        tr5 = np.trace(U[b, ...].dot(U[a, ...]).dot(E[c, ...])[:nocc, :nocc])
+        tr6 = np.trace(U[a, ...].dot(U[b, ...]).dot(E[c, ...])[:nocc, :nocc])
+        tl = tl1 + tl2 + tl3
+        tr = tr1 + tr2 + tr3 + tr4 + tr5 + tr6
+        hyperpolarizability[r, a] = -2 * (tl - tr)
+
+ref_static = np.array([
+    [ 0.00000001,   0.00000000,   0.22843772],
+    [ 0.00000000,   0.00000000, -25.35476040],
+    [ 0.00000000,   0.00000000, -10.84023375],
+    [ 0.00000000,   0.00000000,   0.00000000],
+    [ 0.22843772,   0.00000000,   0.00000000],
+    [ 0.00000000, -25.35476040,   0.00000000]
+])
+assert np.allclose(ref_static, hyperpolarizability, rtol=0.0, atol=1.0e-3)
+print('\nFirst dipole hyperpolarizability (static):')
+print(hyperpolarizability)
+
+print('Setting up for second-harmonic generation (SHG) calculation...')
+f1 = 0.0773178
+f2 = 2 * f1
+
+print('\nForming response vectors for {} a.u.'.format(f1))
+helper1 = helper_CPHF(mol)
+helper1.solve_dynamic_direct(omega=f1)
+helper1.form_polarizability()
+print(helper1.polar)
+print('\nForming response vectors for {} a.u.'.format(f2))
+helper2 = helper_CPHF(mol)
+helper2.solve_dynamic_direct(omega=f2)
+helper2.form_polarizability()
+print(helper2.polar)
+
+rspvecs1 = helper1.x
+rspvecs2 = helper2.x
+
+# repack response vectors to [norb, norb]
+U1 = np.zeros_like(integrals_mo)
+U2 = np.zeros_like(integrals_mo)
+for i in range(ncomp):
+    U1[i, :nocc, nocc:] = rspvecs1[i][nov:].reshape(nocc, nvir)
+    U1[i, nocc:, :nocc] = rspvecs1[i][:nov].reshape(nocc, nvir).T
+    U2[i, :nocc, nocc:] = rspvecs2[i][nov:].reshape(nocc, nvir)
+    U2[i, nocc:, :nocc] = rspvecs2[i][:nov].reshape(nocc, nvir).T
+
+G1 = np.empty_like(U1)
```

Comment:
Which equations does the following belong to?

Reply:
Everything with a 1 corresponds to the given frequency, and everything with a 2 corresponds to the doubled frequency, as this is for beta(-2w;w,w). I'll be more clear about that.

Reply:
Ok, let me know if the comment I added isn't clear enough.